crypto/store/loader_file.c

```diff
@@ -0,0 +1,1027 @@
+                    goto p12_end;
+                }
+                if (!PKCS12_verify_mac(p12, pass, strlen(pass))) {
+                    STOREerr(STORE_F_TRY_DECODE_PKCS12,
+                             STORE_R_ERROR_VERIFYING_PKCS12_MAC);
+                    goto p12_end;
+                }
+            }
+
+            if (PKCS12_parse(p12, pass, &pkey, &cert, &chain)) {
+                if ((ctx = sk_STORE_INFO_new_null()) != NULL
+                    && sk_STORE_INFO_push(ctx, STORE_INFO_new_PKEY(pkey)) != 0
+                    && sk_STORE_INFO_push(ctx, STORE_INFO_new_CERT(cert)) != 0
+                    && (ok = 1))
+                    while(sk_X509_num(chain) > 0) {
+                        X509 *ca = sk_X509_value(chain, 0);
+
+                        if (sk_STORE_INFO_push(ctx, STORE_INFO_new_CERT(ca))
+                            == 0) {
+                            ok = 0;
+                            break;
+                        }
+                        (void)sk_X509_shift(chain);
+                    }
+                if (!ok) {
+                    sk_STORE_INFO_pop_free(ctx, STORE_INFO_free);
+                    EVP_PKEY_free(pkey);
+                    X509_free(cert);
+                    sk_X509_pop_free(chain, X509_free);
+                    ctx = NULL;
+                }
+                *pctx = ctx;
+            }
+        }
+     p12_end:
+        PKCS12_free(p12);
+        if (!ok)
+            /* The caller takes care of destroying my ctx */
+            return NULL;
+    }
+
+    if (ctx != NULL)
+        store_info = sk_STORE_INFO_shift(ctx);
+
+    return store_info;
+}
+static int eof_PKCS12(void *ctx_)
+{
+    STACK_OF(STORE_INFO) *ctx = ctx_;
+
+    return ctx == NULL || sk_STORE_INFO_num(ctx) == 0;
+}
+static void destroy_ctx_PKCS12(void **pctx)
+{
+    STACK_OF(STORE_INFO) *ctx = *pctx;
+
+    sk_STORE_INFO_pop_free(ctx, STORE_INFO_free);
+    *pctx = NULL;
+}
+static STORE_FILE_HANDLER PKCS12_handler = {
+    "PKCS12",
+    try_decode_PKCS12,
+    eof_PKCS12,
+    destroy_ctx_PKCS12,
+    1                            /* repeatable */
+};
+
+static STORE_INFO *try_decode_PKCS8Encrypted(const char *pem_name,
+                                             const char *pem_header,
+                                             const unsigned char *blob,
+                                             size_t len, void **pctx,
+                                             const UI_METHOD *ui_method,
+                                             void *ui_data)
+{
+    if (pem_name == NULL
+        || strcmp(pem_name, PEM_STRING_PKCS8) == 0) {
+        X509_SIG *p8 = d2i_X509_SIG(NULL, &blob, len);
+
+        if (p8) {
+            char kbuf[PEM_BUFSIZE];
+            char *pass = NULL;
+            const X509_ALGOR *dalg = NULL;
+            const ASN1_OCTET_STRING *doct = NULL;
+            STORE_INFO *store_info = NULL;
+            BUF_MEM *mem = OPENSSL_zalloc(sizeof(*mem));
```

Comment:
> I think it's better to have things be opaque as much as possible.

Agreed, calling OPENSSL_zalloc() directly implies knowledge about what BUF_MEM_new() does. If at some point in the future we change BUF_MEM_new() to work slightly differently then this could introduce subtle bugs here.